### PR TITLE
Add RzenEmbed model

### PR DESCRIPTION
**Pull request by deep9539**
- [x] I have filled out the ModelMeta object to the extent possible
- [x] I have ensured that my model can be loaded using
  - [x] `mteb.get_model(model_name, revision)` and
  - [x] `mteb.get_model_meta(model_name, revision)`
- [x] I have tested the implementation works on a representative set of tasks.
- [x] The model is public, i.e., is available either as an API or the weights are publicly available to download

* Flickr30kI2TRetrieval: 0.92021
* MSCOCOI2TRetrieval: 0.7217

Thanks @nehal309 for the help here.

### CODE REVIEW

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+    open_weights=True,
+    max_tokens=32768,
+    reference="https://huggingface.co/qihoo360/RzenEmbed",
+    similarity_fn_name=ScoringFunction.COSINE,
+    framework=["PyTorch", "Transformers"],
+    use_instructions=True,
+    public_training_code=None,
+    public_training_data=None,
+    training_datasets=None,
+)
```
Comment (ayush1298) [inline mteb/models/model_implementations/rzen_embed_model.py:379]: Can you add a citation, and also check if the paper as anything on datasets?

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
+from __future__ import annotations
+
+import logging
+import math
+from typing import TYPE_CHECKING, Any
+
+import numpy as np
+import torch
+from PIL import Image
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py:9]: This should be inside model import

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+        image_obj = to_pil_image(image.cpu())
+    elif isinstance(image, dict) and "bytes" in image:
+        image_obj = Image.open(BytesIO(image["bytes"]))
+    elif isinstance(image, str):
+        image_obj = Image.open(image)
+    else:
+        raise TypeError(
+            f"Unrecognized or unsupported image input format: {type(image)}"
+        )
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py:89]:
```suggestion
```

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+
+        min_image_tokens = kwargs.get("min_image_tokens", 256)
+        max_image_tokens = kwargs.get("max_image_tokens", 1280)
+        min_video_tokens = kwargs.get("min_video_tokens", 160)
+        max_video_tokens = kwargs.get("max_video_tokens", 180)
+        max_length = kwargs.get("max_length", 2000)
+        fps = kwargs.get("fps", 2.0)
+        max_frames = kwargs.get("max_frames", 64)
+        num_frames = kwargs.get("num_frames", None)
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py:118]: You can explicitly list this in init

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+
+        self.device = device or ("cuda" if torch.cuda.is_available() else "cpu")
+        self.max_length = max_length
+        self.normalize = True
+        self.use_instructions = True
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py:123]:
```suggestion
```

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+                input_str += "<|vision_start|><|image_pad|><|vision_end|>" * len(
+                    processed_images
+                )
+            else:
+                processed_images = None
+
+        if text is not None:
+            input_str += text
+
+        system_instruction = instruction if instruction else self.default_instruction
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py:194]: In their implementation they always use default instruction and append task instruction to input_str https://huggingface.co/qihoo360/RzenEmbed/blob/main/rzen_embed_inference.py#L202

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+        all_embeddings = []
+
+        for batch in tqdm(inputs, desc="RzenEmbed Processing"):
+            input_texts, batch_images = self._process_batch_elements(batch, instruction)
+
+            inputs_tokenized = (self.video_processor if has_video else self.processor)(
+                text=input_texts,
+                images=batch_images,
+                padding=True,
+                return_tensors="pt",
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py:292]: They also truncate texts https://huggingface.co/qihoo360/RzenEmbed/blob/main/rzen_embed_inference.py#L173

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+    name="qihoo360/RzenEmbed",
+    model_type=["dense"],
+    languages=["eng-Latn", "zho-Hans"],
+    revision="ea95c339008f9639420fcb33ba2ba606ea94be78",
+    release_date="2025-11-06",
+    modalities=["image", "video", "text"],
+    n_parameters=8_291_375_616,
+    memory_usage_mb=16584,
+    embed_dim=3584,
+    license="apache-2.0",
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py:369]: Where do you found license?

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+                        image_embeds = image_embeds.last_hidden_state
+                    elif (
+                        hasattr(image_embeds, "image_embeds")
+                        and image_embeds.image_embeds is not None
+                    ):
+                        image_embeds = image_embeds.image_embeds
+                    else:
+                        image_embeds = image_embeds[0]
+
+                image_embeds = self.model.model.visual.merger(image_embeds)
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py:324]: Why this needed? I can't find in source implementation

Comment (deep9539) [reply mteb/models/model_implementations/rzen_embed_model.py]: good catch

File: mteb/models/model_implementations/rzen_embed_model.py (hunk 8 above)
Comment (deep9539) [reply mteb/models/model_implementations/rzen_embed_model.py]: sorry, it's mit from github

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+rzen_embed = ModelMeta(
+    loader=RzenEmbedWrapper,
+    name="qihoo360/RzenEmbed",
+    model_type=["dense"],
+    languages=["eng-Latn", "zho-Hans"],
+    revision="ea95c339008f9639420fcb33ba2ba606ea94be78",
+    release_date="2025-11-06",
+    modalities=["image", "video", "text"],
+    n_parameters=8_291_375_616,
+    memory_usage_mb=16584,
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py:391]: You should fill `n_embedding_parameters`

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+    license="mit",
+    open_weights=True,
+    max_tokens=32768,
+    reference="https://huggingface.co/qihoo360/RzenEmbed",
+    similarity_fn_name=ScoringFunction.COSINE,
+    framework=["PyTorch", "Transformers"],
+    use_instructions=True,
+    public_training_code=None,
+    public_training_data=None,
+    training_datasets=None,
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py]: Do they report training datasets?

File: mteb/models/model_implementations/rzen_embed_model.py [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+
+            if pixel_values is not None:
+                pixel_values = pixel_values.type(self.model.model.visual.get_dtype())
+                image_embeds = self.model.model.visual(
+                    pixel_values, grid_thw=inputs_tokenized["image_grid_thw"]
+                )
+
+                if not isinstance(image_embeds, torch.Tensor):
+                    image_embeds = getattr(image_embeds, "last_hidden_state", image_embeds[0])
+
```
Comment (Samoed) [inline mteb/models/model_implementations/rzen_embed_model.py]: I run some investigation with LLM


The wrapper only works on transformers ≥ 5.0.0 — it crashes on most of mteb's supported range (BLOCKING)

mteb pins `transformers>=4.40.0` and this PR adds no extra / version pin. The Qwen2-VL vision API changed twice:

| transformers | `self.model.model` = `Qwen2VLModel` | has `.visual` / `.language_model`? | `visual(pixel_values, grid_thw=…)` returns |
|---|---|---|---|
| **4.40 – 4.51** | text-only decoder (`embed_tokens`, `layers`, `norm`) | **no** | plain `Tensor`, already merged → dim 3584 (`return self.merger(hidden_states)`) |
| **4.52 – 4.57.x** | combined wrapper | **yes** | plain `Tensor`, already merged → dim 3584 (`return self.merger(hidden_states)`) |
| **5.0.0+** (installed: 5.16.1) | combined wrapper | **yes** | `BaseModelOutputWithPooling`: `last_hidden_state` = **pre-merger** (dim 1280), `pooler_output` = merged (dim 3584) |

- `BaseModelOutputWithPooling` return landed in **transformers 5.0.0** (verified: absent in v4.57.1, present in v5.0.0).
- The `Qwen2VLModel` refactor adding `.visual` / `.language_model` landed in **4.52** (verified: v4.51.3 has neither, v4.52.4 has both).

PR code:

```python
image_embeds = self.model.model.visual(pixel_values, grid_thw=…)
if not isinstance(image_embeds, torch.Tensor):
    image_embeds = getattr(image_embeds, "last_hidden_state", image_embeds[0])
image_embeds = self.model.model.visual.merger(image_embeds)   # <- unconditional
```

Behavior per version:

- **< 4.52:** `self.model.model.language_model` doesn't exist → `AttributeError` at the `embed_tokens` line, on *every* input (text included). Dead on arrival.
- **4.52 – 4.57.x:** text-only tasks work. Any **image/video** input: `visual()` returns an already-merged Tensor (3584); the `isinstance` guard is `True`, so the extraction branch is skipped, then `.merger()` is called **again** on the merged tensor. `PatchMerger` does `self.ln_q(x)` with `ln_q = LayerNorm(context_dim=1280)` and `.view(-1, 1280*4)` → `RuntimeError` shape mismatch. Hard crash.
- **5.0.0+:** `visual()` returns `BaseModelOutputWithPooling`, guard is `False`, it grabs the pre-merger `last_hidden_state` (1280) and merges once → **correct**. Only range it works on; happens to be what's installed here.

The `isinstance(image_embeds, torch.Tensor)` guard is backwards for the old API: it skips extraction exactly when `visual()` returns a bare tensor (4.x), then double-merges. It only helps in the 5.x case that was already fine.

Comment (deep9539) [reply mteb/models/model_implementations/rzen_embed_model.py]: Ohh nice catch! Yes, you are right -- adding if condition to make it backward compatible. One needs to use mteb[qwen-vl]

Comment (deep9539) [reply mteb/models/model_implementations/rzen_embed_model.py]: (Resolving this, but please feel free to reopen)

**Comment by Samoed** [comment]
> Pending: Verification and running on the dataset remaining.

Did you verify results?

**Comment by deep9539** [comment]
> > Pending: Verification and running on the dataset remaining.
> 
> Did you verify results?

Yes, modified description

**Comment by Samoed** [comment]
How these scores compared to paper?

**Comment by deep9539** [comment]
> How these scores compared to paper?

| Task Name | Average main_score |
|---|---|
| Vidore2BioMedicalLecturesRetrieval | 0.61130 |
| Vidore2ESGReportsHLRetrieval | 0.67807 |
| Vidore2ESGReportsRetrieval | 0.54877 |
| Vidore2EconomicsReportsRetrieval | 0.59857 |
| **Overall Average** | **0.60918** |

Paper reported average of 60.7 under VDRv2 (NDCG@5), ours is 60.9.

**Review by Samoed — APPROVED** [review]
